Introduce implicit last expression results

hamishknight opened the pull request:
Allow implicitly treating the last expression of a function/closure as the result, in addition to the last expression of an if/switch/do expression branch. This is gated behind the experimental feature `ImplicitLastExprResults`.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please build toolchain

hamishknight commented:
FWIW I still need to add some extra tests

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please build toolchain

hamishknight commented:
https://github.com/apple/llvm-project/pull/8051

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/8051

@swift-ci please build toolchain

hamishknight commented:
https://github.com/apple/llvm-project/pull/8051

@swift-ci please test macOS

hamishknight commented:
https://github.com/apple/llvm-project/pull/8051

@swift-ci please build toolchain Windows

hamishknight commented:
https://github.com/apple/llvm-project/pull/8051

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/8051

@swift-ci please test macOS

hamishknight commented:
https://github.com/apple/llvm-project/pull/8051

@swift-ci please test macOS

airspeedswift commented:
https://github.com/apple/llvm-project/pull/8051

@swift-ci please build toolchain

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please build toolchain

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please build toolchain

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

airspeedswift commented:
@swift-ci please build toolchain

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please build toolchain

hamishknight commented:
@swift-ci please test macOS

hamishknight committed:
Allow implicit last expression results for `if`/`switch` expressions

Allow implicitly treating the last expression of
the branch as the result, behind the experimental
feature `ImplicitLastExprResults`.

hamishknight committed:
[CS] Generalize implied result handling

Track the implied result exprs in the constraint
system, and allow arbitrary propagation of
implied results down if/switch expression
branches. This is required for allowing implied
results in non-single-expression closures.

hamishknight committed:
[IDE] Simplify `isImplicitSingleExpressionReturn`

Use the generalized implied result logic, and
rename to `isImpliedResult` since that's really
what we're querying here, and it needs to handle
implicit-last-exprs if enabled.

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test Windows

hamishknight commented:
Source compat failures are the same as on main

ahoppen reviewed (commented):
Code completion changes in https://github.com/apple/swift/pull/71006/commits/91c01771ef00b72d920da099409f65a29c492420 look good to me

xedin commented on include/swift/Sema/SyntacticElementTarget.h:
```diff
@@ -457,6 +464,11 @@ class SyntacticElementTarget {
     return cast<ExprPattern>(expression.pattern);
   }
 
+  ReturnStmt *getReturnStmt() const {
```
Maybe `getParentReturnStmt()` instead to indicate that we are returning contextual node that wraps the expression?

hamishknight committed:
[CS] Remove `CTP_ImpliedReturnStmt`

Unify with `CTP_ReturnStmt`, and have the
SyntacticElementTarget carry the ReturnStmt for
regular type-checking, which we can use to record
implied returns.

hamishknight committed:
[CS] Account for implicit `return`s in closures

Avoid defaulting to Void if we have an implicit
return from which we can infer the result type.

hamishknight committed:
Allow implicit last expressions for functions and closures

Gated behind the experimental feature
`ImplicitLastExprResults`.

hamishknight commented:
@swift-ci please test
